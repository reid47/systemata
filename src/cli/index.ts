#!/usr/bin/env node
import * as path from 'path';
import * as fs from 'fs';
import { build } from './build';
import { version } from '../../package.json';
import { parseArgs, knownCommands } from './parse-args';
import { usage } from './usage';

const error = (message: string) => {
  console.error(`${message}\n\nUse '-h' or '--help' to show usage information.`);
  process.exit(1);
};

const { args, command } = parseArgs(process.argv.slice(2));

if (args.help) {
  console.log(usage);
  process.exit(0);
}

if (args.version) {
  console.log(version);
  process.exit(0);
}

if (command === 'init') {
  const configFile = args['-c'] || args['--config'];

  if (!configFile || typeof configFile !== 'string') {
    error("No config file specified. Use '-c' or '--config' to pass a path to a system config file.");
  }

  if (fs.existsSync(path.resolve(process.cwd(), configFile as string))) {
    error(`File '${configFile}' already exists. Please specify a file name that doesn't exist.`);
  }
}

if (command === 'build') {
  const configFile = args['-c'] || args['--config'];

  if (!configFile || typeof configFile !== 'string') {
    error("No config file specified. Use '-c' or '--config' to pass a path to a system config file.");
  }

  let config;
  try {
    config = require(path.resolve(process.cwd(), configFile as string));
  } catch (err) {
    error(`Error loading config file: ${err.message}`);
  }

  try {
    build(config);
    process.exit(0);
  } catch (err) {
    error(err);
  }
}

const validCommands = Object.keys(knownCommands).join(', ');

if (!command) {
  error(`No command specified. Command should be one of: ${validCommands}`);
}

error(`Invalid command specified: ${command}\nCommand should be one of: ${validCommands}`);
