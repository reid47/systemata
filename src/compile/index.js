import fs from 'fs';
import path from 'path';
import * as scss from './to-scss';

export default function compile(config) {
  const ext = scss.extension;
  fs.writeFileSync(
    path.resolve(__dirname, '../../ignored/wow' + ext),
    scss.compile(config),
    'utf8'
  );
}
