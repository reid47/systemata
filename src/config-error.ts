export class ConfigurationError extends Error {
  constructor(configPath: string, message: string) {
    const fullMessage = `Invalid configuration at "${configPath}": ${message}`;
    super(fullMessage);
  }
}
