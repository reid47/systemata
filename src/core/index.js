const { CssBuilder } = require('../css-builder');

function generate(config) {
  const { typography } = config;

  const cssBuilder = new CssBuilder();

  Object.keys(typography.sizes).forEach(typeSize => {
    cssBuilder.addRule(typeSize, {
      'font-size': typography.sizes[typeSize]
    });
  });

  console.log(JSON.stringify(config, null, 2));
  console.log(cssBuilder.toCss());
}

module.exports = { generate };
