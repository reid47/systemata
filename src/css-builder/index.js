class CssBuilder {
  constructor() {
    this.rules = {};
  }

  addRule(selector, styles) {
    this.rules[selector] = {
      ...this.rules[selector],
      ...styles
    };
  }

  toCss() {
    const lines = [];

    Object.keys(this.rules).forEach(selector => {
      lines.push(`${selector} {`);
      Object.keys(this.rules[selector]).forEach(property => {
        const value = this.rules[selector][property];
        lines.push(`  ${property}: ${value};`);
      });
      lines.push('}\n');
    });

    return lines.join('\n');
  }
}

module.exports = { CssBuilder };
