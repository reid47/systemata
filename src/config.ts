import * as path from 'path';
import {
  NamespaceConfig,
  Config,
  ColorConfig,
  PropertyMappingConfig,
  SpacingConfig,
  OutputConfig,
  SystemConfig,
  OutputFileConfig,
  DevelopmentConfig
} from './types';
import { version as systemataVersion } from '../package.json';
import { parseSemVer } from './utils';
import { ConfigurationError } from './config-error';

function resolveSystem(systemConfig: any): SystemConfig {
  systemConfig = systemConfig || {};

  const version = systemConfig.version || '0.0.1';
  if (!parseSemVer(version).valid) {
    throw new ConfigurationError(
      'system.version',
      `'${version}' is not a three-part version string (e.g. '0.1.2').`
    );
  }

  return {
    name: systemConfig.name || 'untitled-system',
    version,
    description: systemConfig.description || ''
  };
}

function resolveDevelopment(devConfig: any): DevelopmentConfig {
  devConfig = devConfig || {};

  return {
    serverPort: devConfig.serverPort || 4700
  };
}

function resolveOutput(outputConfig: any, configFileDir: string): OutputConfig {
  outputConfig = outputConfig || {};

  const files: OutputFileConfig[] = [];

  (outputConfig.files || []).forEach((file: any, index: number) => {
    file = file || {};

    const fileName = file.fileName || '';
    if (!fileName) {
      throw new ConfigurationError(`output.files[${index}].fileName`, 'File name must be specified.');
    }

    files.push({
      format: file.format || 'css',
      fileName
    });
  });

  return {
    buildDirectory: path.resolve(configFileDir, outputConfig.buildDirectory || './systemata-build'),
    archiveDirectory: path.resolve(configFileDir, outputConfig.archiveDirectory || './systemata-archive'),
    files
  };
}

function resolveNamespace(namespaceConfig: any): NamespaceConfig {
  namespaceConfig = namespaceConfig || {};

  return {
    prefix: namespaceConfig.prefix || false,
    parentClass: namespaceConfig.parentClass || false,
    peerClass: namespaceConfig.peerClass || false
  };
}

function resolvePropertyMapping(mappingsConfig: any): PropertyMappingConfig {
  mappingsConfig = (mappingsConfig || {}) as object;

  const defaultMappings: PropertyMappingConfig = {
    'background-color': 'bg',
    'border-color': 'brc',
    color: 'c',
    fill: 'fill',
    margin: 'm',
    'margin-bottom': 'mb',
    'margin-left': 'ml',
    'margin-right': 'mr',
    'margin-top': 'mt',
    'margin-top,margin-bottom': 'mv',
    'margin-left,margin-right': 'mh',
    padding: 'p',
    'padding-bottom': 'pb',
    'padding-left': 'pl',
    'padding-right': 'pr',
    'padding-top': 'pt',
    'padding-top,padding-bottom': 'pv',
    'padding-left,padding-right': 'ph'
  };

  return { ...defaultMappings, ...mappingsConfig };
}

function resolveColors(colorConfig: any): ColorConfig {
  if (!colorConfig) return {};
  return colorConfig as ColorConfig;
}

function resolveSpacing(spacingConfig: any): SpacingConfig {
  if (!spacingConfig) return {};
  return spacingConfig as SpacingConfig;
}

export function resolveConfig(config: any, configFilePath: string): Config {
  config = config || {};

  const configFileDir = path.dirname(configFilePath);

  return {
    __configFilePath: configFilePath,
    __systemataVersion: systemataVersion,
    system: resolveSystem(config.system),
    development: resolveDevelopment(config.development),
    namespace: resolveNamespace(config.namespace),
    propertyMapping: resolvePropertyMapping(config.propertyMapping),
    output: resolveOutput(config.output, configFileDir),
    color: resolveColors(config.color),
    spacing: resolveSpacing(config.spacing)
  };
}
