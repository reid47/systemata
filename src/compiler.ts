import { resolveConfig } from './config';
import { Config } from './types';
import { generateCss } from './generate-css';
import { generateSass } from './generate-sass';
import { generateLess } from './generate-less';
import { buildRuleMap, buildVariableMap } from './rule-map';
import { generateCssVariables } from './generate-css-variables';

export class Compiler {
  config: Config;

  constructor(config: any) {
    this.config = resolveConfig(config);
  }

  toString() {
    const { output } = this.config;

    const variableMap = buildVariableMap(this.config);
    const ruleMap = buildRuleMap(this.config, variableMap);

    switch (output.format) {
      case 'css':
        return generateCss(ruleMap);
      case 'sass':
        return generateSass(variableMap, ruleMap);
      case 'less':
        return generateLess(variableMap, ruleMap);
      case 'css-variables':
        return generateCssVariables(variableMap, ruleMap);
    }

    throw new Error('Unsupported output configuration');
  }
}
