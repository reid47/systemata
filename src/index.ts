import { Compiler } from './compiler';
import { OutputOptions } from './types';

export function generate(config: any, output: any): string {
  return new Compiler(config, output).toString();
}
