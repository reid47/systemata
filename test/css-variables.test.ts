import { generate } from '../src/index';

const output = { format: 'css-variables' };

test('color styles', () => {
  const config = {
    output,
    color: {
      primary: { value: '#f00' },
      secondary: { value: '#0f0' }
    }
  };

  expect(generate(config)).toMatchSnapshot();
});

test('spacing styles', () => {
  const config = {
    output,
    spacing: {
      small: { value: '8px' },
      large: { value: '64px' }
    }
  };

  expect(generate(config)).toMatchSnapshot();
});
