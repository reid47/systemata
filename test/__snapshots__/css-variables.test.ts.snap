// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`color styles 1`] = `
":root {
  --color-primary: #f00;
  --color-secondary: #0f0;
}
.bg-primary { background-color: var(--color-primary) }
.bg-secondary { background-color: var(--color-secondary) }
.brc-primary { border-color: var(--color-primary) }
.brc-secondary { border-color: var(--color-secondary) }
.c-primary { color: var(--color-primary) }
.c-secondary { color: var(--color-secondary) }
.fill-primary { fill: var(--color-primary) }
.fill-secondary { fill: var(--color-secondary) }"
`;

exports[`spacing styles 1`] = `
":root {
  --spacing-small: 8px;
  --spacing-large: 64px;
}
.m-small { margin: var(--spacing-small) }
.m-large { margin: var(--spacing-large) }
.mt-small { margin-top: var(--spacing-small) }
.mt-large { margin-top: var(--spacing-large) }
.mb-small { margin-bottom: var(--spacing-small) }
.mb-large { margin-bottom: var(--spacing-large) }
.ml-small { margin-left: var(--spacing-small) }
.ml-large { margin-left: var(--spacing-large) }
.mr-small { margin-right: var(--spacing-small) }
.mr-large { margin-right: var(--spacing-large) }
.mv-small { margin-top: var(--spacing-small); margin-bottom: var(--spacing-small) }
.mv-large { margin-top: var(--spacing-large); margin-bottom: var(--spacing-large) }
.mh-small { margin-left: var(--spacing-small); margin-right: var(--spacing-small) }
.mh-large { margin-left: var(--spacing-large); margin-right: var(--spacing-large) }
.p-small { padding: var(--spacing-small) }
.p-large { padding: var(--spacing-large) }
.pt-small { padding-top: var(--spacing-small) }
.pt-large { padding-top: var(--spacing-large) }
.pb-small { padding-bottom: var(--spacing-small) }
.pb-large { padding-bottom: var(--spacing-large) }
.pl-small { padding-left: var(--spacing-small) }
.pl-large { padding-left: var(--spacing-large) }
.pr-small { padding-right: var(--spacing-small) }
.pr-large { padding-right: var(--spacing-large) }
.pv-small { padding-top: var(--spacing-small); padding-bottom: var(--spacing-small) }
.pv-large { padding-top: var(--spacing-large); padding-bottom: var(--spacing-large) }
.ph-small { padding-left: var(--spacing-small); padding-right: var(--spacing-small) }
.ph-large { padding-left: var(--spacing-large); padding-right: var(--spacing-large) }"
`;
